mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -57,204 +54,235 @@ pub struct TakerSwapWatcherData {
                 wait_taker_payment,
                 WAIT_CONFIRM_INTERVAL,
             )
             .compat();
         if let Err(err) = wait_f.await {
-            return Ok((Some(WatcherCommand::Finish), vec![
-                WatcherEvent::TakerPaymentWaitConfirmFailed(
-                    ERRL!("!taker_coin.wait_for_confirmations: {}", err).into(),
-                ),
-            ]));
+            Self::change_state(Stopped::from_reason(StopReason::TakerPaymentWaitConfirmFailed(
+                ERRL!("!watcher.wait_for_confirmations: {}", err).into(),
+            )));
         }
 
         let validate_input = WatcherValidatePaymentInput {
-            payment_tx: self.data.taker_payment_hex.clone(),
-            time_lock: self.data.taker_payment_lock as u32,
-            taker_pub: self.data.taker_pub.clone(),
-            maker_pub: self.data.maker_pub.clone(),
-            secret_hash: self.data.secret_hash.clone(),
-            amount: self.data.taker_amount.clone(),
+            payment_tx: watcher_ctx.data.taker_payment_hex.clone(),
+            time_lock: watcher_ctx.data.taker_payment_lock as u32,
+            taker_pub: watcher_ctx.data.taker_pub.clone(),
+            maker_pub: watcher_ctx.data.maker_pub.clone(),
+            secret_hash: watcher_ctx.data.secret_hash.clone(),
+            amount: watcher_ctx.data.taker_amount.clone(),
             try_spv_proof_until: wait_taker_payment,
             confirmations,
         };
 
-        let validated_f = self.taker_coin.watcher_validate_taker_payment(validate_input).compat();
+        let validated_f = watcher_ctx
+            .taker_coin
+            .watcher_validate_taker_payment(validate_input)
+            .compat();
 
         if let Err(e) = validated_f.await {
-            return Ok((Some(WatcherCommand::Finish), vec![
-                WatcherEvent::TakerPaymentValidateFailed(
-                    ERRL!("!taker_coin.watcher_validate_taker_payment: {}", e).into(),
-                ),
-            ]));
+            Self::change_state(Stopped::from_reason(StopReason::TakerPaymentValidateFailed(
+                ERRL!("!watcher.watcher_validate_taker_payment: {}", e).into(),
+            )));
         }
 
-        Ok((Some(WatcherCommand::WaitForTakerPaymentSpend), vec![
-            WatcherEvent::TakerPaymentValidatedAndConfirmed,
-        ]))
+        Self::change_state(WaitForTakerPaymentSpend {})
     }
+}
+
+#[async_trait]
+impl State for WaitForTakerPaymentSpend {
+    type Ctx = WatcherContext;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, watcher_ctx: &mut WatcherContext) -> StateResult<Self::Ctx, Self::Result> {
+        #[cfg(not(test))]
+        {
+            // Sleep for half the locktime to allow the taker to spend the maker payment first
+            let now = now_ms() / 1000;
+            let wait_for_taker_until = watcher_ctx.data.swap_started_at + (watcher_ctx.data.lock_duration / 2);
```

Comment:
I could miss and forgot something, please correct me if I'm wrong:
* Taker waits for Maker to spend his funds till `started_at + (lock_duration * 4) / 5`:
https://github.com/KomodoPlatform/atomicDEX-API/blob/2665257607b3e735993772a899383889281eac9f/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L1225-L1226

So what do you think about sleeping till `started_at + (lock_duration * 4) / 5`?

Reply:
We talked with Artem that the watcher should give some time to the taker before spending the `taker_spends_maker_payment`, so I made them wait for half the locktime. But on second thought after your comment, if the maker spends the taker payment inbetween (locktime/2) and (locktime * 4) / 5 then watchers might react earlier than the taker, so it makes sense to extend the duration until we're sure that takers aren't waiting anymore. What do you think @artemii235?